test/ruby/test_yjit.rb

```diff
@@ -825,6 +825,41 @@ def foo
     RUBY
   end
 
+  def test_code_gc
+    assert_compiles(<<~'RUBY', exits: :any, result: :ok)
+      def compiles(&block)
+        failures = RubyVM::YJIT.runtime_stats[:compilation_failure]
+        block.call
+        failures == RubyVM::YJIT.runtime_stats[:compilation_failure]
+      end
+
+      def add_pages(num_jits)
+        pages = RubyVM::YJIT.runtime_stats[:compiled_page_count]
+        100.times { return false unless eval('compiles { nil.to_i }') }
+        pages.nil? || pages < RubyVM::YJIT.runtime_stats[:compiled_page_count]
+      end
+
+      def code_gc
+        RubyVM::YJIT.simulate_oom! # bump write_pos
+        eval('proc { nil }.call') # trigger code GC
+      end
+
+      return :not_paged unless add_pages(100) # prepare freeable pages
+      code_gc # first code GC
+      return :not_compiled1 unless compiles { nil } # should be JITable again
+
+      code_gc # second code GC
+      return :not_compiled2 unless compiles { nil } # should be JITable again
+
+      code_gc_count = RubyVM::YJIT.runtime_stats[:code_gc_count]
+      return :"code_gc_#{code_gc_count}" if code_gc_count && code_gc_count != 2
+
+      :ok
+    RUBY
+  end
```

Comment:
I wish we had more code GC tests, testing situations where we have a jitted parent function on the stack for example, or using eval to generate a lot of code and jitting it. Could be in another PR if you prefer though.

Reply:
The on-stack case is a good catch; Alan and I wrote a script to test that case, we confirmed that it's working, and then I forgot about including it as a test.

About JITing many functions though, this test was like that at first (use `--jit-exec-mem-size=1` and actually run out of it instead of using `simulate_oom!`), but I changed it a bit because my tolerance of long testing time per test case was not high. However, maybe it's fair and worthwhile to spend more time in GC testing to hopefully test more cases of free page jumps. I can resurrect that case too.

Reply:
added some more cases https://github.com/ruby/ruby/pull/6406/commits/e77c015655cfdd4093a90c837ee562711d427e7b, trying to cover more edge cases than each test's primary interest.